### PR TITLE
Complete Port of our Redis.js onto Satellite

### PR DESCRIPTION
**Pull request by Metropass**
In issue [#1908](https://github.com/Seneca-CDOT/telescope/issues/1908), we discussed about porting Redis onto Satellite for use on other modules.

Everything here should work (in theory), but I need some help on writing tests for Redis.

### CODE REVIEW

File: src/redis.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  try {
+    const { port, host } = new URL(redisUrl);
+    return new redisConstructor(port, host, {
+      password: process.env.REDIS_PASSWORD,
+    });
+  } catch (error) {
+    const message = `Unable to parse port and host from "${redisUrl}"`;
+    logger.error({ error }, message);
+    throw new Error(message);
+    //createError(error, message); Future Proofing when createError is merged.
```
Comment (Metropass) [inline src/redis.js:24]: I left this commented out, until createError is merged. Once that happens, I'll be using that instead of throwing an Error.

**Comment by c3ho** [comment]
What tests are you trying to write?

**Comment by Metropass** [comment]
> What tests are you trying to write?

Just a simple test that makes sure that Redis and MockRedis are working as expected

**Comment by c3ho** [comment]
I'm assuming you're just checking for connectivity then?

**Comment by Metropass** [comment]
> I'm assuming you're just checking for connectivity then? If so, Redis has a PING command, if it is connected should return PONG https://redis.io/commands/PING

Noted, thanks! I'll add that in `test.js`

**Comment by humphd** [comment]
This needs a rebase.

**Comment by Metropass** [comment]
@humphd I might have to move the Redis test onto Telescope instead of Satellite. I need to have Docker running within Telescope to get the tests to pass.

**Comment by Metropass** [comment]
Yup, just tested it. If we try to bring Redis onto our tests, it will continue to run until we explicitly kill it. Should we add a function that will allow us to terminate the process?

File: package.json [hunk 2]
```diff
@@ -21,13 +21,16 @@
   "dependencies": {
     "@elastic/ecs-pino-format": "^1.0.0",
     "@godaddy/terminus": "^4.7.0",
+    "array-from": "^2.1.1",
```
Comment (humphd) [inline package.json:24]: Unused, please remove this.

File: src/redis.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+if (useMockRedis && typeof MockRedis.prototype.info !== 'function') {
+  logger.debug('Shimming MockRedis info() method');
+  MockRedis.prototype.info = () => Promise.resolve('redis_version:999.999.999');
+}
+
+module.exports = {
+  // If callers need to create a new redis instance, they'll use the ctor
+  createRedisClient,
+  // Otherwise they can use this shared instance (most should use this)
+  redis: createRedisClient(),
```
Comment (humphd) [inline src/redis.js:38]: This is why you're having tests not finish, due to open connections.  Let's change the exports:

```js
module.exports = createRedisClient()
```

And in `index.js` let's expose it on Satellite like this:

```js
module.exports.Redis = require('./src/redis')`
```

This way anyone who wants to use Redis in a microservice would do this:

```js
const { Redis }  = require('@senecacdot/satellite');
...
const redis = Redis();
// use redis...
...
// when shutting down, disconnect the redis client to close connections
redis.disconnect();
```

File: src/redis.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+const logger = require('./logger');
+const Redis = require('ioredis');
+const MockRedis = require('ioredis-mock');
+
+// If you need to set the Redis URL, do it in REDIS_URL
+const redisUrl = 'http://redis:6379';
```
Comment (humphd) [inline src/redis.js:6]: This seems wrong.  First, it's using `http://` vs. `redis://`, and second, it's not using an env variable.

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
Has this stalled?

**Comment by humphd** [comment]
@Metropass what's up with this PR?

**Comment by Metropass** [comment]
> @Metropass what's up with this PR?

Had to postpone it for now, at least until I finish with my Cache middleware, and progress further with the CORS Middleware. ETA on that should be around 3 days

**Comment by humphd** [comment]
When this goes in, remind me that we need to use it in the Auth service to do Redis backed sessions.

**Comment by HyperTHD** [comment]
#11 successfully ports redis to Satellite. Closing this PR

**Pull request closed without merging**